docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+  instanceType: t3.large
+  iamInstanceProfile: "control-plane.cluster-api-provider-aws.sigs.k8s.io"
+  sshKeyName: my-ssh-key
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: AWSCluster
+metadata:
+  name: my-awscluster
+  namespace: default
+spec:
+  region: us-east-1
+```
+
+![controlplane-init-1](images/controlplane/controlplane-init-1.png)
+![controlplane-init-2](images/controlplane/controlplane-init-2.png)
+![controlplane-init-3](images/controlplane/controlplane-init-3.png)
+![controlplane-init-4](images/controlplane/controlplane-init-4.png)
+
+##### Scale Up
+
+- Allow scale up a control plane with stacked etcd to only 1, 3, 5, or 7, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
+- However, allow a control plane using an external etcd cluster to scale up to other numbers such as 2 or 4.
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
+- Scale up operations should not be done in conjunction with an upgrade operation.
+- Scale up operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+
+![controlplane-init-6](images/controlplane/controlplane-init-6.png)
+
+##### Scale Down
+
+- Scale down a stacked control plane
+  - For each replica being deleted:
+    - Remove the replica’s etcd member from the etcd cluster
+    - Update the kubeadm generated config map
+- Scale down a control plane with an external etcd
+  - Same as above minus Etcd management
+- Scale down operations should not be done in conjunction with an upgrade operation, this should not impact manual operations for recovery.
+- Scale down operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+- Scale to 0 should be rejected for the initial support
+
+![controlplane-init-7](images/controlplane/controlplane-init-7.png)
+
+##### Delete of the entire ControlPlane (kubectl delete controlplane my-controlplane)
+
+- Completely removing the control plane and issuing a delete on the underlying machines.
+  - Deletion is equivalent to a scale to 1 followed by a deletion of a single replica Control Plane.
+
+##### Cluster upgrade (using create-swap-and-delete)
+
+- Triggered on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
+    - Possible implementations include (not necessary to hammer out exact implementation right now):
```

Comment:
Given that the above is a `Must` and not future work, this proposal should either contain a spec and implementation details, or remove it from here and move it to future work instead / open an issue

Reply:
I'd like to keep the MUST in place as a forcing function to require us to deliver the functionality for v1alpha3, but let's defer actual implementation until later so that we can avoid rushing the design right now and gather further feedback for the UX around this.